pkg/jvscrypto/kmsutil.go

```diff
@@ -93,38 +94,108 @@ func getLabelValue(versionName string) (string, error) {
 	return versionValue, nil
 }
 
-// PublicKeysFor returns a map of a Cloud KMS key version name to the public key
-// PEM for that key version. It only returns keys that are enabled.
-func PublicKeysFor(ctx context.Context, client *kms.KeyManagementClient, parentKey string) (map[string]crypto.PublicKey, error) {
-	it := client.ListCryptoKeyVersions(ctx, &kmspb.ListCryptoKeyVersionsRequest{
-		Parent: parentKey,
-		Filter: "state=ENABLED",
-	})
-
-	result := make(map[string]crypto.PublicKey)
-	for {
-		keyVersion, err := it.Next()
-		if errors.Is(err, iterator.Done) {
-			break
+// CryptoKeyVersionsFor returns the list of cryptoKeyVersions for all the given
+// parent keys.
+func CryptoKeyVersionsFor(ctx context.Context, client *kms.KeyManagementClient, parentKeys []string) ([]string, error) {
+	// Accumulate all the key versions for all provided keys.
+	versionsWorker := worker.New[[]string](0)
+	for _, parentKey := range parentKeys {
+		parentKey := parentKey
+
+		if err := versionsWorker.Do(ctx, func() ([]string, error) {
+			it := client.ListCryptoKeyVersions(ctx, &kmspb.ListCryptoKeyVersionsRequest{
+				Parent: parentKey,
+				Filter: "state=ENABLED",
+			})
+
+			var keyVersions []string
+			for {
+				keyVersion, err := it.Next()
+				if errors.Is(err, iterator.Done) {
+					break
+				}
+				if err != nil {
+					return nil, fmt.Errorf("failed to get key version for %s: %w", parentKey, err)
+				}
+				keyVersions = append(keyVersions, keyVersion.Name)
+			}
+			return keyVersions, nil
+		}); err != nil {
+			return nil, fmt.Errorf("failed to get versions for %s: %w", parentKey, err)
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to get key version: %w", err)
+	}
+
+	versionsResults, err := versionsWorker.Done(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch versions: %w", err)
+	}
+
+	keyVersions := make(map[string]struct{}, len(versionsResults))
+	for _, result := range versionsResults {
+		if result.Error != nil {
+			return nil, err
 		}
 
-		publicKeyResp, err := client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{
-			Name: keyVersion.Name,
-		})
-		if err != nil {
-			return nil, fmt.Errorf("failed to get public key for key version %s: %w", keyVersion.Name, err)
+		for _, v := range result.Value {
+			keyVersions[v] = struct{}{}
 		}
+	}
```

Comment:
This logic is to deduplicate key versions. But I wonder why it's needed?

Reply:
Since `parentKeys` is a slice, there's nothing that prevents a user from giving duplicate parent keys names. That would result in us returning multiple key versions.

It would be the same if we de-duped parentKeys, but this felt more defensive.